File: py/other/RPN.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class RPN:
    def converse(self, values: List[str]) -> List[str]:
        stack, result = [], []
        for value in values:
            if '(' == value:
                stack.append(value)
                continue

            if ')' == value:
                while stack:
                    sign = stack.pop()
                    if '(' == sign:
                        break
                    result.append(sign)
                continue

            if '+' != value and '-' != value and '*' != value and '/' != value:
                result.append(value)
                continue

            while stack:
                sign = stack[-1]
                if '(' == sign or (('*' == value or '/' == value) and ('+' == sign or '-' == sign)):
                    break
                result.append(sign)
                stack.pop()
            stack.append(value)

        while stack:
            result.append(stack.pop())
        return result

    def evaluate(self, values: List[str]) -> int:
        stack = []
        for value in values:
            if '+' != value and '-' != value and '*' != value and '/' != value:
                stack.append(int(value))
                continue

            num1, num2 = stack.pop(), stack.pop()
            if '+' == value:
                stack.append(num2 + num1)
            elif '-' == value:
                stack.append(num2 - num1)
            elif '*' == value:
                stack.append(num2 * num1)
            else:
                stack.append(num2 // num1)
        return stack[0]

File: cpp/other/RPN.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
using namespace std;

class RPN {
public:
    vector<string> converse(vector<string> &values) {
        stack<string> stack;
        vector<string> result(0);
        for (string value : values) {
            if (!value.compare("(")) {
                stack.push(value);
                continue;
            }

            if (!value.compare(")")) {
                while (!stack.empty()) {
                    string sign = stack.top();
                    stack.pop();
                    if (!sign.compare("(")) {
                        break;
                    }
                    result.emplace_back(sign);
                }
                continue;
            }

            if (value.compare("+") && value.compare("-") && value.compare("*") && value.compare("/")) {
                result.emplace_back(value);
                continue;
            }

            while (!stack.empty()) {
                string sign = stack.top();
                if (!sign.compare("(") || ((!value.compare("*") || !value.compare("/")) && (!sign.compare("+") || !sign.compare("-")))) {
                    break;
                }
                result.emplace_back(sign);
                stack.pop();
            }
            stack.push(value);
        }

        while (!stack.empty()) {
            result.emplace_back(stack.top());
            stack.pop();
        }
        return result;
    }

    int evaluate(vector<string> &values) {
        stack<int> stack;
        for (string value : values) {
            if (value.compare("+") && value.compare("-") && value.compare("*") && value.compare("/")) {
                stack.push(stoi(value));
                continue;
            }

            int num1 = stack.top();
            stack.pop();
            int num2 = stack.top();
            stack.pop();
            if (!value.compare("+")) {
                stack.push(num2 + num1);
            } else if (!value.compare("-")) {
                stack.push(num2 - num1);
            } else if (!value.compare("*")) {
                stack.push(num2 * num1);
            } else {
                stack.push(num2 / num1);
            }
        }
        return stack.top();
    }
};
